src/routers/alpha-router/alpha-router.ts

```diff
@@ -171,6 +175,10 @@ export type AlphaRouterParams = {
    * The provider for getting V3 quotes.
    */
   v2QuoteProvider?: IV2QuoteProvider;
+  /**
+   * The provider for getting on chain quotes (V3 or MixedRoute)
+   */
+  mixedRouteQuoteProvider?: IOnChainQuoteProvider<MixedRoute>;
```

Comment:
Chatted w/ sara offline abt this, and yeah I agree that we should not have two instances of the `OnChainQuoteProvider` class. I can combine them and push up an implementation soon! However, we thought that enabling getting onChain v2Routes but only for exactInput (because we don't support exactOut at the contract level for mixedRouteQuoter) was a bit weird, because you'd expect to be able to get on chain v2 quotes for exactOut. Wdyt?

Reply:
Also, we chose to make `mixedRouteQuoteProvider` undefined for chains that don't have V2 liquidity via the code here in AutoRouter:
```
        /// @dev We only explicitly support chains with V2 liquidity for mixedRoutes.
        ///      so by default, mixedRouteQuoteProvider is undefined
        case ChainId.RINKEBY:
        case ChainId.ROPSTEN:
        case ChainId.KOVAN:
        case ChainId.GÖRLI:
        case ChainId.MAINNET:
          this.mixedRouteQuoteProvider = new OnChainQuoteProvider(
```
Which I liked for design choice, since it prevented usage of the mixedQuotesProvider for unsupported chains. If we combine the v3 and mixedRoute ones we could throw the error at runtime when quotes are fetched though

Reply:
> Chatted w/ sara offline abt this, and yeah I agree that we should not have two instances of the `OnChainQuoteProvider` class. I can combine them and push up an implementation soon! However, we thought that enabling getting onChain v2Routes but only for exactInput (because we don't support exactOut at the contract level for mixedRouteQuoter) was a bit weird, because you'd expect to be able to get on chain v2 quotes for exactOut. Wdyt?

Bit odd but I think its ok, just as long as its called out in the docs why this is the case, and you recommend users to use the offchain v2 quote provider instead

Reply:
> Also, we chose to make `mixedRouteQuoteProvider` undefined for chains that don't have V2 liquidity via the code here in AutoRouter:
> 
> ```
>         /// @dev We only explicitly support chains with V2 liquidity for mixedRoutes.
>         ///      so by default, mixedRouteQuoteProvider is undefined
>         case ChainId.RINKEBY:
>         case ChainId.ROPSTEN:
>         case ChainId.KOVAN:
>         case ChainId.GÖRLI:
>         case ChainId.MAINNET:
>           this.mixedRouteQuoteProvider = new OnChainQuoteProvider(
> ```
> 
> Which I liked for design choice, since it prevented usage of the mixedQuotesProvider for unsupported chains. If we combine the v3 and mixedRoute ones we can throw the `chain not supported` error at runtime when quotes are fetched instead of at construction now

I don't think we need to return an error- we can just not return them the mixed routes even if they explicitly requested it. 

Its similar to V2- if someone specifies V2 as a protocol on a non-V2-supported chain we don't error out we just ignore it and find them the best route in the other protocols they provided. If they specify V2 only on a non-V2 chain, we just return route not found

Reply:
Got it, that pattern makes sense and I think that'll solve some issues I was running into thinking through the implementation